File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/CardsOnMapAdapter.java
package za.co.woolworths.financial.services.android.ui.adapters;

import android.app.Activity;
import android.os.Parcelable;
import androidx.viewpager.widget.PagerAdapter;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.AbsoluteSizeSpan;
import android.view.View;
import android.view.ViewGroup;

import com.awfs.coordination.R;

import java.util.List;

import za.co.woolworths.financial.services.android.models.dto.StoreDetails;
import za.co.woolworths.financial.services.android.ui.views.WTextView;
import za.co.woolworths.financial.services.android.util.WFormatter;

import static android.text.Spanned.SPAN_INCLUSIVE_INCLUSIVE;

/**
 * Created by W7099877 on 10/10/2016.
 */

public class CardsOnMapAdapter extends PagerAdapter {
    public Activity mContext;
    public List<StoreDetails> storeDetailsList;

    public CardsOnMapAdapter(Activity context, List<StoreDetails> storeDetailsList) {
        this.mContext = context;
        this.storeDetailsList = storeDetailsList;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {

        container.removeView((View) object);
    }


    @Override
    public Object instantiateItem(ViewGroup container, final int position) {

        View cView = mContext.getLayoutInflater().inflate(R.layout.store_nearby_item, container, false);
        WTextView storeName = (WTextView) cView.findViewById(R.id.storeName);
        WTextView storeOfferings = (WTextView) cView.findViewById(R.id.offerings);
        WTextView storeDistance = (WTextView) cView.findViewById(R.id.distance);
        WTextView storeAddress = (WTextView) cView.findViewById(R.id.storeAddress);
        WTextView storeTimeing = (WTextView) cView.findViewById(R.id.timeing);
        storeName.setText(storeDetailsList.get(position).name);
        if (!TextUtils.isEmpty(storeDetailsList.get(position).address))
            storeAddress.setText(storeDetailsList.get(position).address);

        int mKmDistance = mContext.getResources().getDimensionPixelSize(R.dimen.distance_km);
        SpannableString ssDistance = new SpannableString(WFormatter.formatMeter(storeDetailsList.get(position).distance));
        SpannableString mSpanKm = new SpannableString(mContext.getResources().getString(R.string.distance_in_km));
        mSpanKm.setSpan(new AbsoluteSizeSpan(mKmDistance), 0, mSpanKm.length(), SPAN_INCLUSIVE_INCLUSIVE);
        CharSequence mDistancekM = TextUtils.concat(ssDistance, "\n", mSpanKm);
        storeDistance.setText(mDistancekM);

        if (storeDetailsList.get(position).offerings != null)
            storeOfferings.setText(WFormatter.formatOfferingString(storeDetailsList.get(position).offerings));
        if (storeDetailsList.get(position).times != null) {
            try {
                String mHour = WFormatter.formatOpenUntilTime(storeDetailsList.get(position).times.get(0).hours);
                storeTimeing.setText("Open until " + mHour);
            } catch (ArrayIndexOutOfBoundsException ignored) {
            }
        }
        container.addView(cView);
        return cView;
    }

    @Override
    public int getCount() {
        return storeDetailsList.size();
    }

    @Override
    public float getPageWidth(int position) {
        return 0.92f;
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view.equals(object);
    }

    @Override
    public void restoreState(Parcelable state, ClassLoader loader) {
        super.restoreState(state, loader);
    }

    @Override
    public Parcelable saveState() {
        return null;
    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/CardsOnMapAdapter.kt
package za.co.woolworths.financial.services.android.ui.adapters

import android.app.Activity
import android.os.Parcelable
import android.text.SpannableString
import android.text.Spanned
import android.text.TextUtils
import android.text.style.AbsoluteSizeSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.awfs.coordination.R
import com.awfs.coordination.databinding.StoreNearbyItemBinding
import za.co.woolworths.financial.services.android.models.dto.StoreDetails
import za.co.woolworths.financial.services.android.util.WFormatter


class CardsOnMapAdapter(context: Activity, storeDetailsList: List<StoreDetails>) :
    PagerAdapter() {
    private var mContext: Activity
    private var storeDetailsList: List<StoreDetails>

    init {
        mContext = context
        this.storeDetailsList = storeDetailsList
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as? View)
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {

        val binding:StoreNearbyItemBinding =
            StoreNearbyItemBinding.inflate(LayoutInflater.from(container.context), container, false)
        binding.storeName.text = storeDetailsList[position].name
        if (!TextUtils.isEmpty(storeDetailsList[position].address)) binding.storeAddress.text =
            storeDetailsList[position].address
        val mKmDistance: Int = mContext.resources.getDimensionPixelSize(R.dimen.distance_km)
        val ssDistance =
            SpannableString(WFormatter.formatMeter(storeDetailsList[position].distance))
        val mSpanKm = SpannableString(mContext.resources.getString(R.string.distance_in_km))
        mSpanKm.setSpan(AbsoluteSizeSpan(mKmDistance),
            0,
            mSpanKm.length,
            Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        val mDistancekM = TextUtils.concat(ssDistance, "\n", mSpanKm)
        binding.distance.text = mDistancekM
        if (storeDetailsList[position].offerings != null) binding.offerings.text = WFormatter.formatOfferingString(
            storeDetailsList[position].offerings)
        if (storeDetailsList[position].times != null) {
            try {
                val mHour =
                    WFormatter.formatOpenUntilTime(storeDetailsList[position].times.getOrNull(0)?.hours)
                binding.timeing.text = mContext.resources.getString(R.string.open_until,mHour)
            } catch (ignored: ArrayIndexOutOfBoundsException) {
                // Do Nothing
            }
        }
        container.addView(binding.root)
        return binding.root
    }

    override fun getCount(): Int {
        return storeDetailsList.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view == `object`
    }

    override fun restoreState(state: Parcelable?, loader: ClassLoader?) {
        super.restoreState(state, loader)
    }

    override fun saveState(): Parcelable? {
        return null
    }
}